lazy serialization of case indices and transactions

snopoke opened the pull request:
Started working on this idea. I think this will make a big reduction in load on the shard databases from UCRs.

Tested on staging:
- [x] form submission
- [x] view form + view case + raw doc
- [x] process form to ES
- [x] process case to ES
- [x] basic case UCR
- [x] case UCR which accesses lazy property

snopoke committed:
lazy serialization of case indices and transactions

snopoke committed:
custom ES serializer

snopoke committed:
remove --stop

emord reviewed (approved):
Nice. I think there's still a lot of usage for indicies, but it should help for actions

snopoke committed:
fix serializer

snopoke committed:
evaluate lazy args

snopoke committed:
serialize with Django encoder

snopoke committed:
use 'many' feature of serializers

snopoke committed:
use custom json serializer

snopoke committed:
fix eval lazy

snopoke committed:
make pillow with only UCR processor

snopoke committed:
make xform_ids and case_attachments lazy

snopoke committed:
use object identify for cache

snopoke committed:
fix num expected queries

snopoke committed:
eval lazy args

snopoke committed:
special case for OrderedDict

snopoke committed:
correctly filter case action diffs

Prior to the diff fix for OrderedDict (993c99028d2) the diff
for case actions was a 'type' diff since the couch version
was a dict and the SQL version was an OrderedDict

After the fix the diff is a per field diff.

snopoke committed:
make form models lazy

snopoke committed:
rename

snopoke committed:
Merge branch 'master' into sk/lazy-case-json

millerdev commented on .travis.yml:
```diff
@@ -39,7 +39,7 @@ script:
       bash scripts/test-make-requirements.sh
     fi
-  - "scripts/docker test --noinput --stop --verbosity=2 --divide-depth=1 --with-timing --threshold=10"
+  - "scripts/docker test --noinput --verbosity=2 --divide-depth=1 --with-timing --threshold=10"
```
Is removing `--stop` an intentional part of this PR?

snopoke replied:
not permanently

millerdev commented on corehq/apps/tzmigration/timezonemigration.py:
```diff
@@ -47,20 +50,24 @@ def _json_diff(obj1, obj2, path, track_list_indices=True):
+    elif isinstance(obj1, dict):
+        if not isinstance(obj2, dict):
+            # special case to deal with OrderedDicts
+            yield FormJsonDiff('type', path, obj1, obj2)
```
In Python 3.6 and later all dicts preserve ordering. Will that matter here?

snopoke replied:
That's fine and won't be an issue here. This was only an issue when there's a mix of `dict` and `OrderedDict`

snopoke committed:
use commcare json serializer in more places

millerdev commented on corehq/elastic.py:
```diff
@@ -33,6 +37,26 @@
+    def dumps(self, data):
+        # don't serialize strings
+        if isinstance(data, six.string_types):
+            return data
```
Does this mean that a string containing JSON-deserializable content will become a different value when deserialized? 

For example:
```py
>>> ejson = ESJSONSerializer()
>>> ejson.loads(ejson.dumps("{}"))
{}

>>> ejson.loads(ejson.dumps("123"))
123

>>> ejson.loads(ejson.dumps("{name}"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 7, in loads
__main__.SerializationError: ('{name}', ValueError('Expecting property name: line 1 column 2 (char 1)',))

>>> json.loads(json.dumps("{name}"))  # Real json dumps/loads for comparison
u'{name}'
```

Does that matter?

snopoke replied:
This is taken from the serializer included in the ES client:  https://github.com/elastic/elasticsearch-py/blob/91cd918477ab0cfcc3a7236b7f651d8254045748/elasticsearch/serializer.py#L40

I'll include a comment to indicate that.

snopoke committed:
[ci skip] docstring

snopoke committed:
spelling

snopoke committed:
add back stop

snopoke committed:
Merge branch 'master' into sk/lazy-case-json

snopoke committed:
test using correct db connection

snopoke committed:
check correct DB

snopoke committed:
[ci skip] future imports

snopoke commented:
putting this on staging

snopoke commented:
I think this is GTG, going to merge master and let build go

snopoke committed:
Merge branch 'master' into sk/lazy-case-json

snopoke committed:
Merge branch 'sk/missing-domain' into sk/lazy-case-json

emord commented on custom/icds_reports/ucr/data_sources/infrastructure_form.json:
```diff
@@ -12,7 +12,8 @@
     "icds-cas",
-    "cas-lab"
+    "cas-lab",
+    "icds-cas-sandbox"
```
this is probably unintentional addition

calellowitz replied:
This is from the branch he merged in https://github.com/dimagi/commcare-hq/pull/24006/commits/400f9acffe2bb9b25e3fcaa25cf57ba620806a85
That commit is part of master now, so shouldn't be an issue.

emord committed:
Merge branch 'master' into sk/lazy-case-json

emord commented:
@millerdev @snopoke took a swing at fixing the conflicts between this and https://github.com/dimagi/commcare-hq/pull/24111/ to unblock a staging rebuild. not sure that I got the diff.py changes exactly correct. please take a look and let me know if i can help fix any issues

millerdev commented on corehq/apps/couch_sql_migration/diff.py:
```diff
@@ -109,6 +108,8 @@
         # defaulted on SQL
         Ignore('missing', ('indices', '[*]', 'relationship'), old=MISSING, new='child'),
 
+        Ignore(path=('actions', '[*]')),
```
I think this needs to change to

```suggestion
        Ignore(check=is_case_actions),
```

and this will need to be added below

```py
def is_case_actions(old_obj, new_obj, rule, diff):
    return diff.path[0] == "actions"
```

snopoke committed:
Merge branch 'master' into sk/lazy-case-json

snopoke committed:
fix case action diff rule

snopoke committed:
Merge branch 'sk/lazy-case-json' of github.com:dimagi/commcare-hq into sk/lazy-case-json

snopoke committed:
lint
